#1261 add collecting parameter pattern (#1715)
iluwatar/java-design-patterns: collecting-parameter/README.md

---
layout: pattern
title: Collecting parameter
folder: collecting-parameter
permalink: /patterns/collecting-parameter/

> categories: Structural

Reviewer: Add a tag as mentioned in the Wiki setion.

Reviewer 2: I'd suggest the category `Idiom` and the tag `Decoupling`

---

## Intent
A Collecting Parameter is an object that you pass to methods in order to collect information from those methods. This pattern is often coupled with Composed Method

> In the CollectingParameter idiom a collection (list, map, etc.) is passed repeatedly as a parameter to a method which adds items to the collection.

Reviewer 2: ```suggestion
In the Collecting Parameter idiom, a collection (list, map, etc.) is passed repeatedly as a parameter to a method that adds items to the collection.
```

## Explanation

Real world example

> Consider that You have a single bulky method that accumulates information to a local variable.
> A Collecting Parameter may also be passed to methods on multiple objects
> Collecting parameter pattern lets you make the accumulating method into some simple method.

Wikipedia says

**Programmatic Example**

> Consider that we need to collect information of a group of apples.

Reviewer 2: ```suggestion
Consider that we need to collect information about a group of apples.
```

First we have class `Apple` with weight and name information

```java
public class Apple {
  int weight;
  String type;

  public Apple(int weight, String type) {
    this.weight = weight;
    this.type = type;
  }
}
```

If we do not use collecting parameter pattern, we should use a bulky method to get the information of the apples.

```java
public class App {

  public static void main(String[] args) {
    List<Apple> apples = new ArrayList<>();

    apples.add(new Apple(10, "a"));
    apples.add(new Apple(20, "b"));
    apples.add(new Apple(15, "c"));

    StringBuilder result = new StringBuilder();
    for(Apple iter : apples) {
      result.append("apple weight: ");
      result.append(iter.weight);
      result.append(" \n");
    }
  }
}
```

> Some times the way we want to accumulate the information in the object might be very complicated, or we want to collect information from different kinds of objects .We can use collecting parameter pattern to simplify it.

Reviewer 2: ```suggestion
Sometimes the way we want to accumulate the information in the object might be very complicated, or we want to collect information from different kinds of objects. We can use the collecting parameter pattern to simplify it.
```

    collectType(apples);
    collectWeight(apples);
  }

  /**
   * collect the Apple weight information by throwing weightInfo to apples.
   *
   * @param apples the apple list need to collect information from
   */
  public static void collectWeight(List<Apple> apples) {
    List<String> weightInfo = new ArrayList<>();
    for (Apple apple : apples) {
      apple.writeWeightTo(weightInfo);
    }
    for (String message : weightInfo) {
      LOGGER.info(message);
    }
  }

  /**
   * collect the Apple type information by throwing typeInfo to apples.
   *
   * @param apples the apple list need to collect information from
   */
  public static void collectType(List<Apple> apples) {
    List<String> typeInfo = new ArrayList<>();
    for (Apple apple : apples) {
      apple.writeTypeTo(typeInfo);
    }
    for (String message : typeInfo) {
      LOGGER.info(message);
    }
  }
}

```

## Class diagram
![alt text](./etc/collectingparameter.urm.png "Adapter class diagram")

## Applicability
Use the collecting parameter pattern when

* You have a single bulky method that accumulates information to a local variable
## Consequences:
Collecting parameter pattern can:

> *	Helps transform bulky methods into smaller, simpler methods.
> *   make resulting code run faster.

Reviewer: Correct the formatting and provide more references on how can it make the code run faster would really help us understand.

```suggestion
* Helps transform bulky methods into smaller, simpler methods.
* Make the resulting code run faster.
```
